### ISSUE TITLE
train_textcat.py:evaluate ignores non-positive examples

### ISSUE THREAD

**Issue by fhamborg** [explosion/spacy-transformers#74]
I was wondering, why in `train_textcat.py` the evaluation function ignores documents where the predicted label != the positive label, cf. https://github.com/explosion/spacy-pytorch-transformers/blob/ab132b674c5a91510eb8cc472cdbdf5877d24145/examples/train_textcat.py#L260 

Doesn't this skew the evaluation - in some cases strongly? For instance, if a hypothetical test dataset consists of 99% negative instances and only 1% positive instances? Don't we need to know how the classifier's prediction was on all the non-positive instances in order to be able to calculate the overall precision, recall and F1.

Thanks in advance for your answer :-)

**Comment by pmbaumgartner** [comment]
In the binary case, it's assumed that any prediction that's "non-positive" is classified as negative. In this case, the prediction is `NEGATIVE` if `score < 0.5`. You can see in the code that it is correctly identifying these as either a true negative (`tn`) if the actual label is `NEGATIVE`, or a false negative (`fn`) if the label is actually `POSITIVE`.

**Comment by fhamborg** [comment]
But the check in line 260 is whether the gold label is the positive class. This doesn't tell us anything about the actual prediction, e.g., assume the classifier predicted the positive class but the gold label is actually negative. Because of this check in line 260 and the following `continue`, this misclassification of positive (which should be negative) would not increasing the `fp` counter, which, however, should be the case.

**Comment by pmbaumgartner** [comment]
```python
            for label, score in doc.cats.items():
                if label not in gold:
                    continue
                if label != pos_label:
                    continue
```

That's not how I'm reading it. For each label in `doc.cats` (which could be multiple categories from outside of this sentiment classification task), check whether that label is in the gold labels (for this task). Then, since this is a binary task, check if the label is the positive label.

> assume the classifier predicted the positive class but the gold label is actually negative.

If the classifier predicted the positive class, neither `continue` here would be executed, and that misclassification would be captured in line `264`.

**Comment by pmbaumgartner** [comment]
And just FYI, when I actually use this code in production I usually don't use this at all and use `f1_score` etc from `sklearn.metrics` because I assume they've tested these types of things. Using that might alleviate your concern as well.

**Comment by fhamborg** [comment]
@pmbaumgartner Thanks for clearing that up! :-) It looks I somehow missed the logic of the `for label,score in doc.cats.items():` statement, with which of course the following lines make sense.
